### DISCUSSION TITLE
NNUE 1: Get reliabilty score beside value

### DISCUSSION THREAD

**Discussion by xefoci7612** [official-stockfish/Stockfish#4051] [Ideas]
Currently we have some quite complex heuristic to chose which evaluation to use.

``` C++
  bool useClassical = (pos.this_thread()->depth > 9 || pos.count<ALL_PIECES>() > 7) &&
          abs(eg_value(pos.psq_score())) * 5 > (856 + pos.non_pawn_material() / 64) * (10 + pos.rule50_count());
```

Not only it is complex, possibly is sub-optimal too.

The idea I suggest is the following:

1. In the set of train positions, beside evaluation value at a given depth, add also classical eval score (or diff between classical and target), and also a pre-trained NNUE eval scores (see point 3).

2. When training, beside training for target value, also train to output _useClassic_ signal defined as:

                        useClassic  = classic eval better than NNUE eval

where better it means closer to target searched value than NNUE. It comes natural to define a loss on it using the differences to target value of classical and NNUE output.

3. Maybe train an already trained net keeping frozen the weights, and only training the part regarding the new _useClassic_ signal. If this is not possible than you may want to compute the _useClassic_ loss not against the position evaluation of the current net under training, for which eval scores are still weak, but against a pre-trained known good NNUE.

4. At inference time, just get the _useClassic_ out of NNUE to decide which evaluation to use

**Comment by dsmsgms** [comment]
I don't think classic eval is better, it is just faster. You are suggesting to make this decision based on net output. How will that more efficient?

**Comment by xefoci7612** [comment]
To give a quantitative answer to your question, one possible way would be to save classical eval and NNUE eval along searched eval when collecting the big set of positions for training. With that info in our hands we could:

1. See if and when classical eval is better than NNUE
2. Eventually infer some heuristic for switching between the 2 (now there is a big complex formula)
3. In case a sensible percent of positions show classical is better, then further work on the ideas of these 2 discussions, otherwise close it, but with an informed decision.
